pkg/istio-agent/health/health_probers.go

```diff
@@ -54,6 +60,63 @@ func (p *ProbeResult) IsHealthy() bool {
 	return *p == Healthy
 }
 
+type GRPCProber struct {
+	Config *v1alpha3.GrpcHealthCheckConfig
+}
+
+var _ Prober = &GRPCProber{}
+
+func NewGRPCProber(cfg *v1alpha3.GrpcHealthCheckConfig) *GRPCProber {
+	return &GRPCProber{
+		Config: cfg,
+	}
+}
+
+func (g *GRPCProber) Probe(timeout time.Duration) (ProbeResult, error) {
+	if g.Config == nil {
+		return Unknown, fmt.Errorf("grpc health check config is nil")
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	conn, err := grpc.DialContext(ctx, net.JoinHostPort("localhost", fmt.Sprintf("%d", g.Config.Port)),
```

Comment:
why localhost rather than the ip address, not all app listens on localhost

Reply:
@hzxuzhonghu referring to it's [implementation in K8s](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/prober/prober.go#L172), looks like the pod IP is used for host, which is equivalent to `localhost` here. Moreover, since we are doing a plain text call here with `insecure.NewCredentials()`, it could only call localhost. WDYT?

PS: Also for reference, [GRPCAction](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L2665-L2676) with no option to configure `Host` vs [HTTPGetAction](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L2315-L2332).

Reply:
Or is the intent here to refer to the loopback IP `127.0.0.1` instead rather than relying on successful resolution of `localhost`?

Reply:
Technically localhost is not equivilent. https://istio.io/latest/blog/2021/upcoming-networking-changes/ explains a bit.

We do this for HTTP https://github.com/istio/istio/blob/9e4ee1054c3cbbf966bff342901b95fea52d8693/pkg/istio-agent/health/health_check.go#L74 which we should probably due.

Note TCP does NOT do this, which is probably a bug and we should also fix it..

TCP does `localhost` implicitly by https://unix.stackexchange.com/a/419881. At https://github.com/istio/istio/blob/bd9e52d05817b03b9811a761a818b2e76d356018/pkg/istio-agent/health/health_probers.go#L157, Host will be empty, which go translates to 0.0.0.0 which the kernel translates to 127.0.0.1

Reply:
@howardjohn @hzxuzhonghu refactored to use common default host across http, tcp and grpc probes. This default host will be used unless overriden via config (in case of http/tcp). Don't think we need override for grpc.